File: src/types/filter.ts
import type { ColumnType } from "@/types/dataset";

export type Operator =
  | "eq"
  | "neq"
  | "contains"
  | "not_contains"
  | "starts_with"
  | "ends_with"
  | "gt"
  | "gte"
  | "lt"
  | "lte"
  | "between"
  | "before"
  | "after"
  | "on"
  | "is_true"
  | "is_false"
  | "is_empty"
  | "is_not_empty"
  | "in";

export type Combinator = "AND" | "OR";

export type FilterRule = {
  id: string;
  column: string;
  operator: Operator;
  value: string;
  value2: string;
};

export type FilterGroup = {
  combinator: Combinator;
  rules: FilterRule[];
};

export type SavedFilter = {
  id: string;
  name: string;
  dataset_id: string;
  dataset_name: string;
  group: FilterGroup;
  created_at: string;
};

export type HistoryEntry = {
  id: string;
  dataset_id: string;
  dataset_name: string;
  group: FilterGroup;
  summary: string;
  matched: number;
  total: number;
  created_at: string;
};

export const OPERATOR_LABELS: Record<Operator, string> = {
  eq: "is equal to",
  neq: "is not equal to",
  contains: "contains",
  not_contains: "does not contain",
  starts_with: "starts with",
  ends_with: "ends with",
  gt: "is greater than",
  gte: "is greater than or equal to",
  lt: "is less than",
  lte: "is less than or equal to",
  between: "is between",
  before: "is before",
  after: "is after",
  on: "is on",
  is_true: "is Yes",
  is_false: "is No",
  is_empty: "is empty",
  is_not_empty: "is not empty",
  in: "is one of",
};

export const OPERATORS_BY_TYPE: Record<ColumnType, Operator[]> = {
  number: ["eq", "neq", "gt", "gte", "lt", "lte", "between", "is_empty", "is_not_empty"],
  text: [
    "eq",
    "neq",
    "contains",
    "not_contains",
    "starts_with",
    "ends_with",
    "in",
    "is_empty",
    "is_not_empty",
  ],
  date: ["on", "before", "after", "between", "is_empty", "is_not_empty"],
  boolean: ["is_true", "is_false", "is_empty", "is_not_empty"],
  empty: ["is_empty", "is_not_empty"],
};

/** Operators that need no value input at all. */
export const NO_VALUE_OPERATORS: Operator[] = ["is_true", "is_false", "is_empty", "is_not_empty"];

File: backend/app/schemas/filter.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

Operator = Literal[
    "eq",
    "neq",
    "contains",
    "not_contains",
    "starts_with",
    "ends_with",
    "gt",
    "gte",
    "lt",
    "lte",
    "between",
    "before",
    "after",
    "on",
    "is_true",
    "is_false",
    "is_empty",
    "is_not_empty",
    "in",
]


class FilterRule(BaseModel):
    id: str = ""
    column: str
    operator: Operator
    value: str = ""
    value2: str = ""


class FilterGroup(BaseModel):
    combinator: Literal["AND", "OR"] = "AND"
    rules: list[FilterRule] = []


class ApplyFilterRequest(BaseModel):
    dataset_id: str
    group: FilterGroup


class ApplyFilterResponse(BaseModel):
    matched: int
    total: int
    summary: str
    rows: list[dict[str, Any]]


class SavedFilterCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    dataset_id: str
    group: FilterGroup


class SavedFilterOut(BaseModel):
    id: str
    name: str
    dataset_id: str
    dataset_name: str
    group: FilterGroup
    created_at: datetime


class HistoryOut(BaseModel):
    id: str
    dataset_id: str
    dataset_name: str
    group: FilterGroup
    summary: str
    matched: int
    total: int
    created_at: datetime
